include/misc/math_extras_impl.h

```diff
@@ -0,0 +1,187 @@
+}
+
+static inline bool u64_add_overflow(u64_t a, u64_t b, u64_t *result)
+{
+	return __builtin_add_overflow(a, b, result);
+}
+
+static inline bool size_add_overflow(size_t a, size_t b, size_t *result)
+{
+	return __builtin_add_overflow(a, b, result);
+}
+#else /* !use_builtin(__builtin_add_overflow) */
+static inline bool u32_add_overflow(u32_t a, u32_t b, u32_t *result)
+{
+	u32_t c = a + b;
+
+	*result = c;
+
+	return c < a;
+}
+
+static inline bool u64_add_overflow(u64_t a, u64_t b, u64_t *result)
+{
+	u64_t c = a + b;
+
+	*result = c;
+
+	return c < a;
+}
+
+static inline bool size_add_overflow(size_t a, size_t b, size_t *result)
+{
+	size_t c = a + b;
+
+	*result = c;
+
+	return c < a;
+}
+#endif /* use_builtin(__builtin_add_overflow) */
+
+#if use_builtin(__builtin_mul_overflow)
+static inline bool u32_mul_overflow(u32_t a, u32_t b, u32_t *result)
+{
+	return __builtin_mul_overflow(a, b, result);
+}
+
+static inline bool u64_mul_overflow(u64_t a, u64_t b, u64_t *result)
+{
+	return __builtin_mul_overflow(a, b, result);
+}
+
+static inline bool size_mul_overflow(size_t a, size_t b, size_t *result)
+{
+	return __builtin_mul_overflow(a, b, result);
+}
+#else /* !use_builtin(__builtin_mul_overflow) */
+static inline bool u32_mul_overflow(u32_t a, u32_t b, u32_t *result)
+{
+	u32_t c = a * b;
+
+	*result = c;
+
+	return a > 0 && c / a != b;
+}
+
+static inline bool u64_mul_overflow(u64_t a, u64_t b, u64_t *result)
+{
+	u64_t c = a * b;
+
+	*result = c;
+
+	return a > 0 && c / a != b;
+}
+
+static inline bool size_mul_overflow(size_t a, size_t b, size_t *result)
+{
+	size_t c = a * b;
+
+	*result = c;
+
+	return a > 0 && c / a != b;
+}
+#endif /* use_builtin(__builtin_mul_overflow) */
+
+#if use_builtin(__builtin_clz)
+static inline int u32_count_leading_zeros(u32_t x)
+{
+	return x == 0 ? 32 : __builtin_clz(x);
+}
+#else /* !use_builtin(__builtin_clz) */
+static inline int u32_count_leading_zeros(u32_t x)
+{
+	int b = 0;
+
+	while (b < 32 && (x >> 31) == 0) {
+		b++, x <<= 1;
+	}
+
+	return b;
+}
+#endif /* use_builtin(__builtin_clz) */
+
+#if use_builtin(__builtin_clzll)
+static inline int u64_count_leading_zeros(u64_t x)
+{
+	return x == 0 ? 64 : __builtin_clzll(x);
+}
+#else /* !use_builtin(__builtin_clzll) */
+static inline int u64_count_leading_zeros(u64_t x)
+{
+	if (x == (u32_t)x) {
+		return 32 + u32_count_leading_zeros((u32_t)x);
+	} else {
+		return u32_count_leading_zeros(x >> 32);
+	}
+}
+#endif /* use_builtin(__builtin_clzll) */
+
+#if use_builtin(__builtin_ctz)
+static inline int u32_count_trailing_zeros(u32_t x)
+{
+	return x == 0 ? 32 : __builtin_ctz(x);
+}
+#else /* !use_builtin(__builtin_ctz) */
+static inline int u32_count_trailing_zeros(u32_t x)
+{
+	int b = 0;
+
+	while (b < 32 && (x & 1) == 0) {
+		b++, x >>= 1;
+	}
+
+	return b;
+}
+#endif /* use_builtin(__builtin_ctz) */
+
+#if use_builtin(__builtin_ctzll)
+static inline int u64_count_trailing_zeros(u64_t x)
+{
+	return x == 0 ? 64 : __builtin_ctzll(x);
```

Comment:
Genuine confusion here.  Why are we making the semantics (for an argument of zero) of our own "count_trailing_zeros" function...  **DIFFERENT** than the "ctz" builtin that we're supposed to be wrapping?  I mean... the whole point to this API is to wrap the builtin!  It should do the same thing, right?

Reply:
Here's how I look at it: The purpose of *this patch* is indeed to wrap compiler builtins and remove them from the general source code. However, looking beyond this PR, the purpose of the API in `math_extras.h` is to provide a set of commonly useful little math routines. For users of the API, it is not interesting whether they are implemented in portable C or with compiler builtins.

The x86 BSF/BSR instructions produce an undefined result for x=0, and GCC turned that into full-blown nasal demons C undefined behavior in their documentation of the ctz/clz builtins. Other instruction set architectures also have clz/ctz instructions, and they all agree that x=0 should return the size of the register. Intel even added LZCNT and TZCNT instructions that do the same.

So since there is a meaningful return value for x=0, I thought it would be better to return that instead of exposing users to nasal demons in a pretty surprising way.

Clang understands what's going on. Here's its armv7 code:

```asm
_u32_count_leading_zeros:
	clz	r0, r0
	bx	lr

_u32_count_trailing_zeros:
	rbit	r0, r0
	clz	r0, r0
	bx	lr
```

Reply:
Seems like this needs an explanatory comment then about the inconsistent edge case, and maybe a `#ifdef __i386` or whatnot for clarity.   While it's nice that clang optimizes this correctly on ARM, I can all but guarantee you that many of our toolchains won't.

Reply:
Unfortunately, `__builtin_ctz(0)` is undefined behavior on *all* architectures, thanks to the way it is defined and documented.

This means that if your code contains a line with `__builtin_ctz(x)` anywhere, the compiler is permitted to assume that x != 0 anywhere it can prove that this line has been or will be executed:

```c
  int first_missile = __builtin_ctz(x);

  if (x == 0) {
      disarm_missiles();
  }

  launch_missiles();
```

Here, the compiler is permitted to remove the conditional. It doesn't matter which instruction it ends up using for the builtin.

Many people find this infuriating, and understandably so.

Reply:
Then... maybe __builtin_ctz() is just problematic and a better solution would be arch-level inline assembly wrappers?  I mean... if the tool is broken why are we using it?

Reply:
@andyross either way, isn't this patch a step in the right direction, in that it gives a single helper entrypoint such that if you want to replace it with an assembly stub later for a specific arch, you can? Meanwhile, using this instead of open-coding __builtin_foo() in the code is a clear incremental improvement.

Reply:
Fair enough.  I'll remove the -1.  Can we at least get a big comment on the usage here explaining that extra test and the fact that it's (hopefully) optimized away for the benefit of people like me would might be tempted to come in and "fix" it later?